### DISCUSSION TITLE
Support for code completion?

### DISCUSSION THREAD

**Discussion by TomzBench** [asomers/mockall#301] [Questions]
I'm not sure if the issue is with rust in general, or my IDE setup.

But for some reason I do not get any code completion support for dotting into my mock objects. That makes it very difficult to use.  Is there something wrong with my IDE? Or is it expected to setup mock objects with out completion?

**Comment by asomers** [comment]
Probably because whatever is generating your completions doesn't expand proc macros.  Does your IDE's completion work with other proc-macro-generated code?  If not, then there's nothing Mockall can do.  If so, then maybe Mockall can be made to work.  But it would probably have to be on a "patches welcome" basis, since I don't use your IDE.

**Reply by TomzBench** [reply]
I'm guessing it's my IDE setup !

I'm using NVIM with YouCompleteMe.  YouCompleteMe uses rust-analyzer. There is extensive rust-analyzer documentation but it's not clear how to enable it via VIM config.

**Reply by asomers** [reply]
Wait, are you saying that ycm isn't working at all with neovim?  It's not just on Mockall's generated code?  Or what part of rust-analyzer can you not enable?

**Reply by TomzBench** [reply]
YCM and Neovim is working pretty good and I get pretty quick completions for my code as well as my dependencies and the stdlib.  The only thing that isn't completing is what i'm guessing is the generated code. For example, `mock.expect_foo()` will compile but it doesn't have completion.  I am just using YCM and Rust-Analyzer out of the box with out any configs added for rust. So i'm guessing what ever is required to expand macros is not enabled.

**Reply by TomzBench** [reply]
For anyone who arrives here with YCM and VIM.  

modify ycmd/build.py
```
diff --git a/build.py b/build.py
index 7b7db323..485b05b1 100755
--- a/build.py
+++ b/build.py
@@ -80,7 +80,7 @@ JDTLS_SHA256 = (
   'df9c9b497ce86b1d57756b2292ad0f7bfaa76aed8a4b63a31c589e85018b7993'
 )

-RUST_TOOLCHAIN = 'nightly-2021-04-14'
+RUST_TOOLCHAIN = 'nightly-2021-05-20'
 RUST_ANALYZER_DIR = p.join( DIR_OF_THIRD_PARTY, 'rust-analyzer' )
```

Then download 6-28-2021 rust-analyzer binary and replace /ycmd/third_party/rust-analyzer/bin/rust-alanyzer

**Reply by asomers** [reply]
Is there an issue for this at ycm-core ?

**Reply by TomzBench** [reply]
I was in the YCM gitter and the maintainers i believe are tracking a PR for fixing this. It's not clear when the fix will be available the normal way though.

Only certain YCM users are affected. It depends on the version of YCM you are using.
